### PR TITLE
Edit comments integration

### PR DESCRIPTION
**Pull request by eldiegod**
fixes: #2440 #2720

### CODE REVIEW

**Review by drillprop — CHANGES_REQUESTED** [review]
There were a lot of changes recently, so I'll review your PR after you resolve all the conflicts

File: packages/atlas/src/hooks/useReactionTransactions.ts
Comment (drillprop) [inline packages/atlas/src/hooks/useReactionTransactions.ts]: What's going on here? Why `commentInputProcessing` is an object with this weirdly named property?

File: packages/atlas/src/hooks/useReactionTransactions.ts
Comment (drillprop) [inline packages/atlas/src/hooks/useReactionTransactions.ts]: Let's not put this in `preProcess`, let's move this before `handleTransaction` as we did in other methods.

File: packages/atlas/src/hooks/useReactionTransactions.ts
Comment (drillprop) [inline packages/atlas/src/hooks/useReactionTransactions.ts:30]: Could you explain why we need to set `commentId` in `commentInputProcessing `? This is because there could be multiple processing comments? This is impossible right now, we can't run multiple transactions at once. We don't block them yet, but I guess we will - https://github.com/Joystream/atlas/issues/2541

Also, I think, in that case, this should be named differently.

File: packages/atlas/src/views/viewer/VideoView/CommentsSection.tsx
Comment (drillprop) [inline packages/atlas/src/views/viewer/VideoView/CommentsSection.tsx]: Again, I'm a bit lost here.  
Why do we have `comment-box` property here and what does it mean(also why you didn't use `COMMENT_BOX_ID` above)?
Why `isEditingComment` is an object? This should be named differently. If I understand this correctly we could create an array or set of `editingCommentsIds` and If someone editing a comment we're adding commentId to this array when someone ends editing, we remove this commentId.  The name `isEditingComment` suggests that this is a primitive value, which could be only a boolean. The name of the variable should always inform you what's inside and what you should expect. 

I speak for myself but I highly doubt this will be understandable to someone who sees this code for the first time or revisits this after two weeks.

File: packages/atlas/src/components/_comments/CommentInput/CommentInput.tsx
Comment (drillprop) [inline packages/atlas/src/components/_comments/CommentInput/CommentInput.tsx]: Why do we need `initialValue`? I think `value` should be enough. Can't we just set `value` once the user triggers an edit comment?

File: packages/atlas/src/components/_comments/CommentInput/CommentInput.tsx
Comment (drillprop) [inline packages/atlas/src/components/_comments/CommentInput/CommentInput.tsx]: I think if my previous comment is true, we shouldn't need that effect

File: packages/atlas/src/components/_comments/Comment/Comment.tsx
Comment (drillprop) [inline packages/atlas/src/components/_comments/Comment/Comment.tsx]: What means `highlighted` and not `_higlighted`?

File: packages/atlas/src/views/viewer/VideoView/CommentsSection.tsx
Comment (drillprop) [inline packages/atlas/src/views/viewer/VideoView/CommentsSection.tsx]: Let's create the handleCancelComment function.

File: packages/atlas/src/components/_comments/Comment/Comment.tsx
Comment (drillprop) [reply packages/atlas/src/components/_comments/Comment/Comment.tsx]: I mean this works, but again I doubt that someone would know what's happening when he revisits this effect.

File: packages/atlas/src/views/viewer/VideoView/CommentsSection.tsx
Comment (drillprop) [inline packages/atlas/src/views/viewer/VideoView/CommentsSection.tsx]: I'm not a big fan of creating variables like `_setSomething` and `setSomething`. Couldn't we rename this handler handleCommentBodyChange or smth?

**Review by drillprop — CHANGES_REQUESTED** [review]
Works pretty good 💪🏻  but I have some concerns about naming. I got lost several times reading the code, I think we could improve this in a few places.

File: packages/atlas/src/hooks/useReactionTransactions.ts
Comment (drillprop) [reply packages/atlas/src/hooks/useReactionTransactions.ts:30]: Oh, okay, nvm - I think I understand, why you need this.
Anyway, I sustain my opinion that this should be named differently though, and also I think we should keep an array or even [set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) of processing comments not an object.

File: packages/atlas/src/hooks/useReactionTransactions.ts
Comment (eldiegod) [reply packages/atlas/src/hooks/useReactionTransactions.ts]: discussed below

File: packages/atlas/src/hooks/useReactionTransactions.ts
Comment (eldiegod) [reply packages/atlas/src/hooks/useReactionTransactions.ts:30]: I disagree with the second part, we always use objects for indexing similar things, I don't think we have ever used a `set` probably because it doesn't play very well with `useState` and an array wouldn't allow for indexing by `CommentId` easily; anyways it could be done both ways

File: packages/atlas/src/views/viewer/VideoView/CommentsSection.tsx
Comment (eldiegod) [reply packages/atlas/src/views/viewer/VideoView/CommentsSection.tsx]: yes, it's exactly as you mention but instead of an array is a standard record (obj) for easier indexing.

I renamed the var for clarity

File: packages/atlas/src/components/_comments/CommentInput/CommentInput.tsx
Comment (eldiegod) [reply packages/atlas/src/components/_comments/CommentInput/CommentInput.tsx]: We need `initialValue` because we need to know if the comment being edited has changed from its `initialValue` before being able to publish it again

**Review by drillprop — CHANGES_REQUESTED** [review]
The comments section is now broken, I don't see any comments. I'll review the code later.

File: packages/atlas/src/views/viewer/VideoView/CommentsSection.tsx
Comment (drillprop) [inline packages/atlas/src/views/viewer/VideoView/CommentsSection.tsx:148]: Ok found the issue - we don't see comments, because you remove `return` here

File: packages/atlas/src/hooks/useReactionTransactions.ts
Comment (drillprop) [reply packages/atlas/src/hooks/useReactionTransactions.ts:30]: Actually, we're using the set in OverlayManager to get information on how many overlays are open. And it works pretty well. We're also using a set in the asset store, but this is a different case.

File: packages/atlas/src/components/_comments/CommentInput/CommentInput.tsx
Comment (drillprop) [reply packages/atlas/src/components/_comments/CommentInput/CommentInput.tsx]: I still sustain that you don't need that. You also don't need this awkward `useEffect` below.  
Hear me out:  
1. Remove initialValue and this useEffect
2. When a user clicks edit, we're calling: 
```
          onEditClick={() => setIsEditingComment({ commentId: comment.id, value: true })}
```
We could add here `setCommentInputTextRecord` and set the `comment.text`(basically what you do in this `useEffect`)
3. Instead of passing initialValue we could pass `isEditChanged` or something, which will be condition `commentInputTextRecord[commentId] === comment.text`
4. We're handling this `isEditChanged` inside the CommentInput and show a dialog if is false.

**Review by drillprop — CHANGES_REQUESTED** [review]
1. The performance is really bad now. Writing in the comment is really laggy.
2. You reintroduce this issue:

https://user-images.githubusercontent.com/51168865/168980326-2af7cc93-c0c2-4b77-9866-dac4f8d6ef1e.mov

just FYI, this worked before.

**Comment by drillprop** [comment]
>  The performance is really bad now. Writing in the comment is really laggy.

This could be a bigger problem and could be caused by something else also(not only your changes).  I've noticed that the same is happening on `reactions` on the video where we have a lot of comments(video/1). I'm looking at it right now.

**Comment by eldiegod** [comment]
> > The performance is really bad now. Writing in the comment is really laggy.
> 
> This could be a bigger problem and could be caused by something else also(not only your changes). I've noticed that the same is happening on `reactions` on the video where we have a lot of comments(video/1). I'm looking at it right now.

@drillprop I'm looking for a solution for this but it's because every time an input triggers on change the commentInputTextRecord changes and it rerenders pretty much the whole comment section. 

But yeah the problem is still there regardless of my changes

**Comment by eldiegod** [comment]
@drillprop I think it should all be good now, performance issues apart

File: packages/atlas/src/components/_comments/CommentInput/CommentInput.tsx
Comment (drillprop) [inline packages/atlas/src/components/_comments/CommentInput/CommentInput.tsx]: Now we don't need that change. I would bring back `onChange` which takes the event as a parameter.

File: packages/atlas/src/hooks/useTrackPropChange.ts
Comment (drillprop) [inline packages/atlas/src/hooks/useTrackPropChange.ts:19]: I like that!

File: packages/atlas/src/views/viewer/VideoView/CommentsSection.tsx
Comment (drillprop) [inline packages/atlas/src/views/viewer/VideoView/CommentsSection.tsx:59]: You didn't have to change this to map 😅  I was fine with using an object in this case.

**Review by drillprop — CHANGES_REQUESTED** [review]
Comment input is not cleared once you add a comment

image

File: packages/atlas/src/components/_comments/CommentInput/CommentInput.tsx
Comment (eldiegod) [reply packages/atlas/src/components/_comments/CommentInput/CommentInput.tsx]: fixed

File: packages/atlas/src/hooks/useReactionTransactions.ts
Comment (drillprop) [inline packages/atlas/src/hooks/useReactionTransactions.ts]: Actually, we should use `videoCommentsConnection` here,  the userComments query ignores comments which are replies to another comment.

**Review by drillprop — CHANGES_REQUESTED** [review]
Last thing

File: packages/atlas/src/hooks/useReactionTransactions.ts
Comment (eldiegod) [reply packages/atlas/src/hooks/useReactionTransactions.ts]: fixed!

**Review by drillprop — APPROVED** [review]
Looks good to me, but I'd wait for final approval from Klaudiusz. 
I requested him to review it.

**Review by kdembler — APPROVED** [review]
Looks good to me as well! I found one small bug - if I start editing a comment, click the confirm button but cancel the transaction, the comments still gets highlighted. Let's fix it in some followup though. @eldiegod can you create an issue or a PR with a fix?

**Pull request merged**
